File: src/main/kotlin/com/huawei/cangjie/psi/stubs/elements/CjStructElementType.java
package com.huawei.cangjie.psi.stubs.elements;

import com.huawei.cangjie.name.ClassId;
import com.huawei.cangjie.name.FqName;
import com.huawei.cangjie.psi.CjClass;
import com.huawei.cangjie.psi.CjInterface;
import com.huawei.cangjie.psi.CjStruct;
import com.huawei.cangjie.psi.psiUtil.CjPsiUtilKt;
import com.huawei.cangjie.psi.psiUtil.StubUtils;
import com.huawei.cangjie.psi.stubs.CangJieClassStub;
import com.huawei.cangjie.psi.stubs.CangJieInterfaceStub;
import com.huawei.cangjie.psi.stubs.CangJieStructStub;
import com.huawei.cangjie.psi.stubs.impl.CangJieClassStubImpl;
import com.huawei.cangjie.psi.stubs.impl.CangJieInterfaceStubImpl;
import com.huawei.cangjie.psi.stubs.impl.CangJieStructStubImpl;
import com.huawei.cangjie.psi.stubs.impl.Utils;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.stubs.IndexSink;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import com.intellij.util.io.StringRef;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.List;

public class CjStructElementType extends CjStubElementType<CangJieStructStub, CjStruct> {

    public CjStructElementType(@NotNull @NonNls String debugName) {
        super(debugName, CjStruct.class, CangJieStructStub.class);
    }

    public void indexStub(@NotNull CangJieStructStub stub, @NotNull IndexSink sink) {
        StubIndexService.getInstance().indexStruct(stub, sink);
    }

    @NotNull
    @Override
    public CjStruct createPsi(@NotNull CangJieStructStub stub) {
        return new CjStruct(stub);
    }


    @NotNull
    @Override
    public CjStruct createPsiFromAst(@NotNull ASTNode node) {
        return new CjStruct(node);
    }

    @Override
    public @NotNull CangJieStructStub createStub(@NotNull CjStruct psi, StubElement<? extends PsiElement> parentStub) {
        FqName fqName = CjPsiUtilKt.safeFqNameForLazyResolve(psi);

        List<String> superNames = CjPsiUtilKt.getSuperNames(psi);
        ClassId classId = StubUtils.createNestedClassId(parentStub, psi);
        return new CangJieStructStubImpl(
                CjStubElementTypes.STRUCT, parentStub,
                StringRef.fromString(fqName != null ? fqName.asString() : null), classId,
                StringRef.fromString(psi.getName()),
                Utils.INSTANCE.wrapStrings(superNames),
                psi.isLocal()
        )  ;
    }

    @Override
    public void serialize(@NotNull CangJieStructStub stub, @NotNull StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());

        FqName fqName = stub.getFqName();
        dataStream.writeName(fqName == null ? null : fqName.asString());

        StubUtils.serializeClassId(dataStream, stub.getClassId());


        dataStream.writeBoolean(stub.isLocal());
//        dataStream.writeBoolean(stub.isTopLevel());

        List<String> superNames = stub.getSuperNames();
        dataStream.writeVarInt(superNames.size());
        for (String name : superNames) {
            dataStream.writeName(name);
        }
    }

    @Override
    public @NotNull CangJieStructStub deserialize(@NotNull StubInputStream dataStream, StubElement parentStub) throws IOException {
        StringRef name = dataStream.readName();
        StringRef qualifiedName = dataStream.readName();

        ClassId classId = StubUtils.deserializeClassId(dataStream);


        boolean isLocal = dataStream.readBoolean();
//        bool isTopLevel = dataStream.readBoolean();

        int superCount = dataStream.readVarInt();
        StringRef[] superNames = StringRef.createArray(superCount);
        for (int i = 0; i < superCount; i++) {
            superNames[i] = dataStream.readName();
        }

        return new CangJieStructStubImpl(
                CjStubElementTypes.STRUCT, (StubElement<?>) parentStub, qualifiedName,classId, name, superNames,
                isLocal
        );
    }
}

File: psi/src/main/kotlin/cn/cangnova/cangjie/psi/stubs/elements/CjStructElementType.kt
/*
 * Copyright 2024 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */
package cn.cangnova.cangjie.psi.stubs.elements

import cn.cangnova.cangjie.psi.CjStruct
import cn.cangnova.cangjie.psi.psiUtil.StubUtils.createNestedClassId
import cn.cangnova.cangjie.psi.psiUtil.StubUtils.deserializeClassId
import cn.cangnova.cangjie.psi.psiUtil.StubUtils.serializeClassId
import cn.cangnova.cangjie.psi.psiUtil.getSuperNames
import cn.cangnova.cangjie.psi.psiUtil.safeFqNameForLazyResolve
import cn.cangnova.cangjie.psi.stubs.CangJieStructStub
import cn.cangnova.cangjie.psi.stubs.elements.StubIndexService.Companion.getInstance
import cn.cangnova.cangjie.psi.stubs.impl.CangJieStructStubImpl
import cn.cangnova.cangjie.psi.stubs.impl.Utils.wrapStrings
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.IndexSink
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import com.intellij.util.io.StringRef
import org.jetbrains.annotations.NonNls
import java.io.IOException

class CjStructElementType(debugName: @NonNls String) :
    CjStubElementType<CangJieStructStub, CjStruct>(debugName, CjStruct::class.java, CangJieStructStub::class.java) {
    override fun indexStub(stub: CangJieStructStub, sink: IndexSink) {
        getInstance().indexStruct(stub, sink)
    }

    override fun createPsi(stub: CangJieStructStub): CjStruct {
        return CjStruct(stub)
    }

    override fun createPsiFromAst(node: ASTNode): CjStruct {
        return CjStruct(node)
    }

    override fun createStub(psi: CjStruct, parentStub: StubElement<out PsiElement?>): CangJieStructStub {
        val fqName = psi.safeFqNameForLazyResolve()

        val superNames = psi.getSuperNames()
        val classId = createNestedClassId(parentStub, psi)
        return CangJieStructStubImpl(
            CjStubElementTypes.STRUCT,
            parentStub,
            StringRef.fromString(fqName?.asString()),
            classId,
            StringRef.fromString(psi.name),
            wrapStrings(superNames),
            psi.isLocal,
        )
    }

    @Throws(IOException::class)
    override fun serialize(stub: CangJieStructStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.name)

        val fqName = stub.getFqName()
        dataStream.writeName(fqName?.asString())

        serializeClassId(dataStream, stub.getClassId())

        dataStream.writeBoolean(stub.isLocal())

        //        dataStream.writeBoolean(stub.isTopLevel());
        val superNames = stub.getSuperNames()
        dataStream.writeVarInt(superNames.size)
        for (name in superNames) {
            dataStream.writeName(name)
        }
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>): CangJieStructStub {
        val name = dataStream.readName()
        val qualifiedName = dataStream.readName()

        val classId = deserializeClassId(dataStream)

        val isLocal = dataStream.readBoolean()

        //        bool isTopLevel = dataStream.readBoolean();
        val superCount = dataStream.readVarInt()
        val superNames = StringRef.createArray(superCount)
        for (i in 0..<superCount) {
            superNames[i] = dataStream.readName()
        }

        return CangJieStructStubImpl(
            CjStubElementTypes.STRUCT,
            parentStub,
            qualifiedName,
            classId,
            name,
            superNames,
            isLocal,
        )
    }

    companion object {
        val stubType: CjStructElementType
            get() = CjStubElementTypes.STRUCT
    }
}
